coniferest/session/callback.py

```diff
@@ -50,3 +54,72 @@ def __call__(self, metadata, data, session) -> None:
         self.iteration += 1
         if self.iteration >= self.budget:
             session.terminate()
+
+
+def autonomous_decider_callback(metadata: Label, data: np.ndarray, session: 'Session') -> Label:
+    "Use metadata as already defined labels for making decision."
+    return metadata
+
+
+class AutonomousExperiment:
+    """
+    A callback for performing experiments with labeled data.
+
+    Parameters
+    ----------
+    budget: int
+        Maximum number of iterations to spend in search of anomalies.
+
+    anomalies: int
+        Maximum number of anomalies to search for.
+
+    save_trace: bool
+        Whether to save the full trace. Default value is True.
+    """
+    def __init__(self, budget: int, anomalies: int, save_trace: bool = True):
+        self._anomalies: int = anomalies
+        self._budget: int = budget
+        self._trajectory: List[int] = []
+        self._labels: List[Label] = []
+        self._trace: List[np.ndarray] = []
+        self._save_trace: bool = save_trace
+
+    @property
+    def trajectory(self) -> 'np.ndarray[np.int_]':
+        return np.array(self._trajectory, dtype=np.int_)
+
+    @property
+    def labels(self) -> 'np.ndarray[np.int8]':
+        return np.array(self._labels, dtype=np.int8)
+
+    @property
+    def trace(self) -> List[np.ndarray]:
+        return self._trace
+
+    def __call__(self, metadata: Label, _, session: 'Session') -> None:
+        self._budget -= 1
+        if metadata == Label.A:
+            self._anomalies -= 1
+
+        if self._anomalies == 0 or self._budget == 0:
+            session.terminate()
+
+        self._trajectory.append(session.current)
+        self._labels.append(session.last_decision)
+
+        if not self._save_trace:
+            return
+
+        ordering = np.argsort(session.scores)
+        index, = np.where(ordering == session.current)
+        self._trace.append(ordering[:index[0] + self._anomalies + 1])
+
+
+def make_autonomous_experiment(
+        metadata: 'np.ndarray[Label]',
+        budget: int = None) -> AutonomousExperiment:
```

Comment:
type should be `Optional[int]`

Reply:
You are right. Maybe `int | None`?

Reply:
It was introduced in Python 3.10 https://peps.python.org/pep-0604/